### ISSUE TITLE
Blanks in output file + out of memory issues during translation

### ISSUE THREAD

**Issue by devinbostIL** [OpenNMT/OpenNMT-py#743]
## Issue description

Blows up with error after translate.py ran for a while. When I inspected the output file, I noticed that only some of the lines (maybe ~ 20% of them) had been randomly translated, so many lines were blank. I attempted to re-run translate.py on a smaller input file (about 8000 lines, rather than over 1,000,000 lines), and the output file was completely blank, but I got the same error. 

Interestingly, I did not get this error when I built a model with the same version of PyOpenNMT on a different dataset, but it's possible that something in my conda environment has changed since then.

## Code example

`(pyTorchLanguage) C:\src\pyopennmt\OpenNMT-py>python preprocess.py -train_src C:\src\torchevere-offensive-classifier\training\character\train_src.txt -train_tgt C:\src\torchevere-offensive-classifier\training\character\train_dst.txt -valid_src C:\src\torchevere-offensive-classifier\training\character\val_src.txt -valid_tgt C:\src\torchevere-offensive-classifier\training\character\val_dst.txt -save_data data/character/tc-offense-classifier-character_v1 -src_seq_length 5000 -tgt_seq_length 5000`

`(pyTorchLanguage) C:\src\pyopennmt\OpenNMT-py>python train.py -data data/character/tc-offense-classifier-character_v1 -save_model tc-offense-classifier-character_v2 -gpuid 0 -layers 3 -learning_rate_decay 0.99 -epochs 100 -rnn_size 500`

`(pyTorchLanguage) C:\src\pyopennmt\OpenNMT-py>python translate.py -model C:\src\pyopennmt\OpenNMT-py\tc-offense-classifier-character_v2_acc_98.54_ppl_1.09_e93.pt -src C:\src\torchevere-offensive-classifier\test.txt -output C:\src\torchevere-offensive-classifier\tc-character_v2_e93.txt`

> Loading model parameters.
> average src size 4.5279249898083975 7359
> Traceback (most recent call last):
>   File "translate.py", line 29, in <module>
>     main(opt)
>   File "translate.py", line 18, in main
>     opt.batch_size, opt.attn_debug)
>   File "C:\src\pyopennmt\OpenNMT-py\onmt\translate\Translator.py", line 158, in translate
>     for batch in data_iter:
>   File "C:\Anaconda3\envs\pyTorchLanguage\lib\site-packages\torchtext\data\iterator.py", line 151, in __iter__
>     self.train)
>   File "C:\Anaconda3\envs\pyTorchLanguage\lib\site-packages\torchtext\data\batch.py", line 27, in __init__
>     setattr(self, name, field.process(batch, device=device, train=train))
>   File "C:\Anaconda3\envs\pyTorchLanguage\lib\site-packages\torchtext\data\field.py", line 188, in process
>     tensor = self.numericalize(padded, device=device, train=train)
>   File "C:\Anaconda3\envs\pyTorchLanguage\lib\site-packages\torchtext\data\field.py", line 308, in numericalize
>     arr = self.postprocessing(arr, None, train)
>   File "C:\src\pyopennmt\OpenNMT-py\onmt\io\TextDataset.py", line 221, in make_src
>     src_size = max([t.size(0) for t in data])
>   File "C:\src\pyopennmt\OpenNMT-py\onmt\io\TextDataset.py", line 221, in <listcomp>
>     src_size = max([t.size(0) for t in data])
> RuntimeError: invalid argument 2: dimension 0 out of range of 0D tensor at c:\anaconda2\conda-bld\pytorch_1519501749874\work\torch\lib\th\generic/THTensor.c:24

## System Info
`(pyTorchLanguage) C:\src\pyopennmt>python collect_env.py`

> Collecting environment information...
> PyTorch version: 0.3.1.post2
> Is debug build: No
> CUDA used to build PyTorch: 9.0
> 
> OS: Microsoft Windows 10 Pro
> GCC version: (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 5.4.0
> CMake version: version 3.10.3
> 
> Python version: 3.6
> Is CUDA available: Yes
> CUDA runtime version: 7.5.17
> GPU models and configuration: Could not collect
> Nvidia driver version: Could not collect
> cuDNN version: Could not collect
> 
> Versions of relevant libraries:
> [pip] Could not collect
> [conda] Could not collect

> CuDNN is v7.0.
> Nvidia driver is v. 397.64.
> GPU model is: Titan Xp
> pip 10.0.1 from `c:\anaconda3\envs\pyTorchLanguage\lib\site-packages\pip (python 3.6)`
> conda 4.5.4

I believe (if I remember correctly) that PyTorch was installed via `pip` (I don't remember exactly how) after pulling the latest source of PyOpenNMT from Github.

Any ideas?

**Comment by devinbostIL** [comment]
BTW, I accidentally posted this as a PyTorch issue: https://github.com/pytorch/pytorch/issues/7982
and they referred me back to here.

**Comment by devinbostIL** [comment]
I created a new (clean) conda environment and reinstalled PyTorch using the latest production build for Windows on CUDA 9.0 and reinstalled PyOpenNMT, and then I got this error:

`(pyTorchOffensive) C:\src\pyopennmt\OpenNMT-py>python translate.py -model C:\src\pyopennmt\OpenNMT-py\tc-offense-classifier-character_v2_acc_98.54_ppl_1.09_e93.pt -src C:\src\torchevere-offensive-classifier\test.txt -output C:\src\torchevere-offensive-classifier\tc-character_v2_e93.txt`

> Loading model parameters.
> average src size 4.5279249898083975 7359
> C:\Anaconda3\envs\pyTorchOffensive\lib\site-packages\torchtext\data\field.py:321: UserWarning: volatile was removed and now has no effect. Use `with torch.no_grad():` instead.
>   return Variable(arr, volatile=not train), lengths
> C:\Anaconda3\envs\pyTorchOffensive\lib\site-packages\torchtext\data\field.py:322: UserWarning: volatile was removed and now has no effect. Use `with torch.no_grad():` instead.
>   return Variable(arr, volatile=not train)
> Traceback (most recent call last):
>   File "translate.py", line 29, in <module>
>     main(opt)
>   File "translate.py", line 18, in main
>     opt.batch_size, opt.attn_debug)
>   File "C:\src\pyopennmt\OpenNMT-py\onmt\translate\Translator.py", line 158, in translate
>     for batch in data_iter:
>   File "C:\Anaconda3\envs\pyTorchOffensive\lib\site-packages\torchtext\data\iterator.py", line 151, in __iter__
>     self.train)
>   File "C:\Anaconda3\envs\pyTorchOffensive\lib\site-packages\torchtext\data\batch.py", line 27, in __init__
>     setattr(self, name, field.process(batch, device=device, train=train))
>   File "C:\Anaconda3\envs\pyTorchOffensive\lib\site-packages\torchtext\data\field.py", line 188, in process
>     tensor = self.numericalize(padded, device=device, train=train)
>   File "C:\Anaconda3\envs\pyTorchOffensive\lib\site-packages\torchtext\data\field.py", line 308, in numericalize
>     arr = self.postprocessing(arr, None, train)
>   File "C:\src\pyopennmt\OpenNMT-py\onmt\io\TextDataset.py", line 222, in make_src
>     src_vocab_size = max([t.max() for t in data]) + 1
>   File "C:\src\pyopennmt\OpenNMT-py\onmt\io\TextDataset.py", line 222, in <listcomp>
>     src_vocab_size = max([t.max() for t in data]) + 1
> RuntimeError: invalid argument 1: tensor must have one dimension at c:\programdata\miniconda3\conda-bld\pytorch_1524549877902\work\aten\src\th\generic/THTensorMath.c:733

**Comment by sebastianGehrmann** [comment]
This sounds like you are running it with pytorch 0.4. Can you try again with 0.3 or this fork: https://github.com/Ubiqus/OpenNMT-py? It will be migrated soon.

**Comment by devinbostIL** [comment]
From the fork, when I tried to install via `pip install -e ` (after uninstalling the previous version of opennmt-py), I got an error message that it couldn't find the directory `onmt/io`
Installing PyTorch 0.3 on Windows requires me to build from source, and that is complicated and a lot of work... But thankfully, I still had a conda environment that was running PyTorch 0.3.

(pyTorchStockMarket) C:\src\pyopennmt>`pip -V`
> pip 10.0.1 from c:\anaconda3\envs\pytorchstockmarket\lib\site-packages\pip (python 3.6)

(pyTorchStockMarket) C:\src\pyopennmt>`conda -V`
> conda 4.5.4

(pyTorchStockMarket) C:\src\pyopennmt>`python`

> Python 3.6.4 |Anaconda, Inc.| (default, Mar 12 2018, 20:20:50) [MSC v.1900 64 bit (AMD64)] on win32
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import torch
> >>> print(torch.__version__)
> **0.3.1.post2**

I then ran:

(pyTorchStockMarket) C:\src\pyopennmt\OpenNMT-py>`python translate.py -model C:\src\pyopennmt\OpenNMT-py\tc-offense-classifier-character_v2_acc_98.54_ppl_1.09_e93.pt -src C:\src\torchevere-offensive-classifier\possibly-non-offensive_student_chats_distinct.csv -output C:\src\torchevere-offensive-classifier\predictions_possibly-non-offensive_student_chats_distinct_tc-offense-classifier-character_v2_e93_attempt2.txt -gpu 0`

It's been running for a while (over an hour on a file with a million lines), so I checked the performance monitor, and the CPU utilization has been running around 7% for the Python process, and the GPU utilization has been around 12% with a spike every once in a while (maybe every 10 minutes).

Is the `translate.py` process single-threaded? I'm wondering if there's anything I can do to speed it up.

**Comment by devinbostIL** [comment]
I still got the same strange issue with missing data in the output file. It also ran for over a day before failing with an out of memory issue:

```
(pyTorchStockMarket) C:\src\pyopennmt\OpenNMT-py>python translate.py -model C:\src\pyopennmt\OpenNMT-py\tc-offense-classifier-character_v2_acc_98.54_ppl_1.09_e93.pt -src C:\src\torchevere-offensive-classifier\possibly-non-offensive_student_chats_distinct.csv -output C:\src\torchevere-offensive-classifier\predictions_possibly-non-offensive_student_chats_distinct_tc-offense-classifier-character_v2_e93_attempt2.txt -gpu 0
Loading model parameters.
average src size 5.998985433147317 1374971
THCudaCheck FAIL file=c:\anaconda2\conda-bld\pytorch_1519501749874\work\torch\lib\thc\generic/THCStorage.cu line=58 error=2 : out of memory
Traceback (most recent call last):
  File "translate.py", line 29, in <module>
    main(opt)
  File "translate.py", line 18, in main
    opt.batch_size, opt.attn_debug)
  File "C:\src\pyopennmt\OpenNMT-py\onmt\translate\Translator.py", line 159, in translate
    batch_data = self.translate_batch(batch, data)
  File "C:\src\pyopennmt\OpenNMT-py\onmt\translate\Translator.py", line 282, in translate_batch
    memory_bank = rvar(memory_bank.data)
  File "C:\src\pyopennmt\OpenNMT-py\onmt\translate\Translator.py", line 256, in rvar
    def rvar(a): return var(a.repeat(1, beam_size, 1))
  File "C:\Anaconda3\envs\pyTorchStockMarket\lib\site-packages\torch\tensor.py", line 290, in repeat
    result.resize_(target_size)
RuntimeError: cuda runtime error (2) : out of memory at c:\anaconda2\conda-bld\pytorch_1519501749874\work\torch\lib\thc\generic/THCStorage.cu:58
```

**Comment by devinbostIL** [comment]
Another thing that seems strange to me is how it's referencing dependencies from the `anaconda2` directory whereas my conda environment clearly should be utilizing only Anaconda 3.

**Comment by devinbostIL** [comment]
It doesn't look like any of the issues mentioned https://github.com/OpenNMT/OpenNMT-py/issues/705 about using PyTorch 0.4 would produce blank lines or additional performance issues.

**Comment by vince62s** [comment]
@devinbostIL  your error with 0.4 / Ubiqus repo (ie onmt/io) is beause we have have breaking changes, you need to reprocess and retrain.

**Comment by devinbostIL** [comment]
Oh, I'll try that. Thanks!

**Comment by devinbostIL** [comment]
@vince62s  when using PyTorch 0.4 in a clean conda environment, when I try installing Ubiqus, I am still getting this error:

```
(pyTorchOffensive) C:\src\pyopennmt\ubiqus\OpenNMT-py>pip install -e .
Obtaining file:///C:/src/pyopennmt/ubiqus/OpenNMT-py
    Complete output from command python setup.py egg_info:
    running egg_info
    creating OpenNMT_py.egg-info
    writing OpenNMT_py.egg-info\PKG-INFO
    writing dependency_links to OpenNMT_py.egg-info\dependency_links.txt
    writing top-level names to OpenNMT_py.egg-info\top_level.txt
    writing manifest file 'OpenNMT_py.egg-info\SOURCES.txt'
    error: package directory 'onmt\io' does not exist

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in C:\src\pyopennmt\ubiqus\OpenNMT-py\
```
Any ideas?

**Comment by vince62s** [comment]
Just pushed a fix, git pull and try again.

**Comment by devinbostIL** [comment]
The installation works now! Thanks!

**Comment by vince62s** [comment]
ok close this and reopen on the fork if you get some questions. thanks.

**Comment by devinbostIL** [comment]
I don't see a tab for issues on the fork Github repo page... 

When I ran:

`(pyTorchOffensive) C:\src\pyopennmt\ubiqus\OpenNMT-py>python train.py -data data/character/tc-offense-classifier-character_v3 -save_model tc-offense-classifier-character_v3 -gpuid 0 -layers 3 -learning_rate_decay 0.99 -epochs 100 -rnn_size 500`

It says:

`train.py: error: unrecognized arguments: -epochs 100`

How do I specify the number of epochs in the new architecture? The closest thing I see in the usage output is `train_steps`.

**Comment by vince62s** [comment]
Yes, sorry.
1) issues tab is on now.
2)  I will publish the release notes on the fork
but yes we switched from epochs to steps.
so you will have train_steps, start_decay_steps, decay_steps
for instance train during 1000000 steps
start to decay at 500000 steps every 100000 steps
(just examples)

**Comment by devinbostIL** [comment]
This issue (with the blanks in the output) was solved by using Ubiqus on PyTorch 0.4. I posted a new issue here where the training crashed at the end of the training process, but I was still able to successfully run `translate.py` from a model file produced during the middle of the training process.

**Comment by WenTsai** [comment]
I got this error on PyTorch 0.4 but not on earlier version with the same `translate commands`.

I could run several `translate` in a time on earlier PyTorch version. However, I got some similar errors as this post mentioned (something like cuda out of memory) on PyTorch 0.4. In fact, the `translate.py` would use lots of memory in GPU. 

I'm not sure why and how to solve this.
